### ISSUE TITLE
Stitch VSCode cannot process object fields interacting with scripts

### ISSUE THREAD

**Issue by stdc105** [bscotch/stitch#117]
For object fields, I mean when inside an object event, declare/assign a variable without using the `var` keyword.

Sample offender code:

`Scripts::ui.gml`
```gml
function ui_create_general() {
  // setup bounding box for an object
  _x1 = 100
  _x2 = display_get_width() - 100
  _y1 = 100
  _y2 = display_get_height() - 100
}
```

`obj_sample::Create_0.gml`

```gml
ui_create_general()

text_y = _y1 + 20 // 1
```

Expected: Stitch doesn't show error on comment 1
Actual: Stitch is reporting an error on comment 1 on text `_y1`: `Undeclared symbol '_y1'`

To be honest I have no idea if this is an issue or not. It's very likely we cannot practically figure out the type if an object field is defined this way, without investing into checking all script calls in all object functions to make sense of an object field scope.

Although it seems like even Stitch is showing an error, `_y1` can live with the type `Any`, which I think it's great -- not the `Any` type is great, but the tool is great to have a fallback and not get into my way. I would be very happy if an `Any` type is showing a different colour so I have some visual indication when I make a typo, instead of an error but having the same colour as other locally defined object fields.

Bug report environment:
Windows 11 Pro 22H2 (22621.1928)
GameMaker BETA IDE v2023.400.0.324 Runtime v2023.400.0.342
VSCode 1.80.1 (user setup)
Stitch VSCode 1.10.1

**Comment by stdc105** [comment]
The same thing can be said with object that is supplied with an optional struct during creation ([manual of `instance_create_layer`](https://beta-manual.yoyogames.com/#t=GameMaker_Language%2FGML_Reference%2FAsset_Management%2FInstances%2Finstance_create_layer.htm)).

```gml
instance_create_layer(0, 0, "Instances", obj_shop, { 
  shop_items: [ 
    { item: "001", price: 100 },
    { item: "002", price: 200 },
  ],
})
```

And in object `obj_shop::Create_0.gml`:

```gml
price_1 = shop_items[0].price // Undeclared symbol 'shop_items'
```

**Comment by adam-coster** [comment]
You can use the `/// @self` JSDoc comment to tell Stitch what the evaluation context of the function will be. Stitch follows the same conventions as Feather in GameMaker. For example:

```js
/// @self Id.Instance.obj_sample
function ui_create_general() {
  // setup bounding box for an object
  _x1 = 100
  _x2 = display_get_width() - 100
  _y1 = 100
  _y2 = display_get_height() - 100
}

// You'd get the same effect out of the following:

/// @context Id.Instance.obj_sample

// -or-

/// @self Asset.GMObject.obj_sample
```

This will cause Stitch to add those variables to the obj_sample object.

Where possible, I'd recommend declaring all instance variables in your object's create event, and then using `@self` (or `@context`, they're aliases) for any functions declared outside of the object's context.

**Comment by stdc105** [comment]
Wow, I didn't know Feather until just now! That would be really helpful for a new project! (I say new project because enabling it using default profile in our current project just got us 1K errors, 3.5K warnings, and 97K suggestions -- I'll stick with Stitch now.)

While using `@self` could provide annotation of one object for a function, in our project functions like `ui_create_general` is used in a lot of places, and there are about 10 functions like this to setup base values of fields. These functions work like constructors in multiple inheritance, and it's a tested pattern to help us reuse code and centralize changes.

Even if it's not impossible to annotate these functions, currently `@self` doesn't work if I put multiple lines of annotations to indicate this function can be executed in any of those objects (it only sets the last annotation as the execution context).

Also, I don't find a good way to annotate struct fields put in the optional struct when calling `instance_create_layer` (see my comment above).

I still believe it's better to just use an `Any` type or a warning is better than showing an error, but enlighten me if you have any good solution to the issues I put here.

**Comment by adam-coster** [comment]
Yeah Feather was in beta for a long time (might still be) and has been changing quite rapidly over the past year. You can configure each Feather rule in GameMaker to be ignored, show up as a warning, or show up as an error, so you can make it a lot quieter if you decide to go through all of those rules.

Stitch has a similar outcome since it uses the same concepts as Feather, where when you use it in an existing project you'll definitely end up with a lot of warnings and errors. I treat these as things to clean up over time whenever I touch a file, and the best way to clean them up is typically to go as up-stream as you can from the problem and add some type annotations.

Stitch also supports the `/// @type {}` annotation for variable declarations, and you can also use `/// @self` before a `with` to fine-tune the context (Feather doesn't support either of those currently).

This issue of using a function as a sort of mixin is something we also have to deal with, and I agree that some sort of way to specify multiple possible contexts would be ideal. Since Feather doesn't support that I'd have to come up with our own approach, and in any case it changes the nature of the problem to make it all significantly more complicated!

What we're doing for this same situation is to refactor these cases so that the functions are instead constructors holding a bag of variables. Instead of sticking those variables right onto the object, you can then keep them in the struct assigned to a single variable in the object. This would be consistent with the design pattern you're trying to use while forcing all declarations to exist on their objects, making all of that completely unambiguous.

For example:

```js
// Scripts:ui.gml
function UiGeneral() constructor {
  // setup bounding box for an object
  _x1 = 100
  _x2 = display_get_width() - 100
  _y1 = 100
  _y2 = display_get_height() - 100
}

// Objects:obj_sample
general = new UiGeneral()

text_y = general._y1 + 20 // 1
```

Obviously this would be a pretty sucky refactor if you're doing this a lot, but the end result is definitely more explicit and easier to understand!

---

Regarding the `instance_create_layer` issue, unfortunately there isn't an obvious way to handle that with Stitch due to the order of operations concerns, unless I implement some logic for that function to make those passed variables available on the target object. I'll make an Issue to add that as a feature, but I'm pretty sure we don't use that function at all so I'm not sure when I'd get around to implementing it.

The only way to solve it with Stitch's current functionality would be to take a completely different approach, where the `shop_items` variable is explicitly declared in the object's Create. You can set the type via Stitch's `/// @type` support, so you could dynamically update the actual values while still having accurate types, but wouldn't be able to use `instance_create_layer` for that.

For example:

```js
/// @param {String} _item
/// @param {Real} _price
function ShopItem (_item, _price){
  item = _item;
  price = _price;
}

// In the object, declare the variable and specify its type annotate the variable type
/// @type {Array<Struct.ShopItem>}
shop_items = [];
```

Another solution I could implement would be to allow somehow adding information about variables to an object just using JSDocs. Or, at least to get rid of the noise, to add an `/// ignore` comment of some sort (Feather has this) so that you could hide specific errors. That latter item is on my ToDo list for once I'm confident that everything is behaving like it's supposed to. For the foreseeable future expect some noise about warnings and errors while the kinks get worked out!

**Comment by stdc105** [comment]
Appreciate the informative response!

Using a real constructor to refactor the mixins (I totally forgot the word) is a great idea! That would also resolve a few namespace problems we are having right now (which require us to call the functions in the correct order -- and that's error prone for sure).

I totally understand that processing the optional struct in `instance_create_layer` isn't easy. Our actual code is more complicated than just making shop items on the fly, and it was clearer for us to understand something is being "passed into" the object than cranking all the interesting code into object creation.

Would something like `/// @type shop_items {Array<Struct.ShopItem>}` serve as a good enough workaround?  Basically declaring a variable in the JSDoc instead of actual code doing that. If you think that's a reasonable idea I can take time in the weekend and look to an implementation -- I have a working plugin environment setup last weekend that I can open our project and see plugin logs in VSCode. It would be more fun to actually contribute come code than just opening issues :)

**Comment by adam-coster** [comment]
Looking through the [JSDoc](https://jsdoc.app/) there are a few ways we could approach this that are consistent with JSDocs. I doubt GameMaker will introduce a feature like this, so as long as we end up with something easy to maintain that doesn't clash with JSDoc standards or likely future Feather features, I'm open to a contribution that adds something like this.

The main source of wonkiness is that JSDocs are meant to describe the following non-comment signifier, or to describe a standalone entity, and since GameMaker objects are created at the *file* level instead of *code* level there's no JSDoc-native way to describe features of a GameMaker object.

One option would be to add a custom JSDoc tag, something like `@object`, meant only for use in Create events of objects. Combined with the [`@property`](https://jsdoc.app/tags-property.html) tag, we could use that to document the current object without it having to apply to a following signifier.

For example, the Create event of your object might start with something like:

```js
/// @object
/// @prop {Array<Struct.ShopItem>} shop_items
```

Since the [`@type`](https://jsdoc.app/tags-type.html) tag, officially, only consists of the `@type {}` components, flipping your example to something like `@type {Array<Struct.ShopItem>} shop_items` would result in syntax similar to other JSDoc tags that we can treat as an extended form of `@type`. This has the advantage of being very concise and pretty intuitive. The main drawback I see is that no scope information is provided about the signifier (e.g. is it a local variable, instance variable, or global?).

Those problems could be solved with a collection of tags that behave the same way as that but that unambiguously indicate the type of variable. For example, `@var {type} name`, `@globalvar {type} name`, and `@instancevar {type} name`. For the latter we could use `@prop`, since that's an official JSDoc tag and it's meant for describing properties of an object/class/struct/whatever.

Do any of those things sound like a good approach? I can definitely see a use case for `@globalvar` for extensions and other things that are difficult to otherwise specify types for.

**Comment by adam-coster** [comment]
I think that, given all of the above, the solution here would be to add the following JSDoc tags to Stitch:

1. `@globalvar {type} name - Optional description` for globals
2. `@var {type} name - Optional description` (probably with `@local` and `@var` as aliases) for locals
3. `@prop {type} name - Optional description` (probably with `@instancevar`, and maybe `@member` and others as aliases

These would actually be pretty simple to add to the existing parser logic, so that's a big plus. At least I think they would be...

**Comment by stdc105** [comment]
I agree that `@globalvar/@var/@prop {type} name` JSDoc extension looks the best. They feel natural in GameMaker context for me and is useful enough for our `global.STUFF` definitions and optional struct instance creations, while preserving the error/warning when weird things are not explicitly annotated.

**Comment by adam-coster** [comment]
Stitch for VSCode v1.11.0 should be available soon, and includes those new JSDoc tags. I added more info to the README about JSDocs in the extension, including these new ones.

**Comment by adam-coster** [comment]
An FYI since it's applicable to this discussion. I just published Stitch 1.13.0, which includes some support for mixin functions:

If you have a global function whose job is to add variables to objects or other contexts, you can now flag it with the `/// @mixin` JSDoc. This does two things:

- Tells Stitch that non-local variable declarations inside the function are allowed (i.e. you won't get "undeclared function" or "it looks like you're adding to global scope" errors)
- Wherever that function is called, Stitch will try to add the instance variables declared inside that function to the current context

That by itself improves editor support for a lot of cases, but we have a lot of cases where we have *other* external functions that reference those same variables. To make it so that Stitch can understand what you're doing, I updated the behavior of `/// @self` to allow you to pass Function types as the self-context. For functions that are constructors, this ends up equivalent to using that constructor's `Struct.Whatever` type. For functions that are *not* constructors, it sets the context to that function's context. For mixins, that context is a struct containing the declared variables.

Here's an example for clarity:

```js
/// @mixin
function init_variables(){
  new_variable = 10;
}

/// @self Function.init_variables
function do_stuff_with_those_variables(){
  var hello = new_variable; // <- works as expected
}
```

All of this is also fully hooked into the references/definitions system, so if you do a "go to definition" on  something declared in a mixin, you'll find yourself inside that mixin. Similarly, if you "find all refs", you'll find that same variable wherever it's used!
